Add debian 9 (stretch) documentation + build fixes

lcaflc opened the pull request:
Add documentation informations about debian 9 (stretch) version of the images.

Please also note that the debian-9 tag is missing in the automated builds produced on docker hub.

Thanks.

lcaflc committed:
Fix debian wheezy typo

lcaflc committed:
Add debian9 (stretch) documentation

Add all documentation informations on debian 9 (stretch) release available really soon.

ChristopherDavenport commented:
Hmmm. I think this may be because the Dockerfile still needs a little work to get correct. Seems `bdist_wheel` is blocking a complete build. 

Here is the [failing build](https://travis-ci.org/AnsibleCheck/ansiblecheck/jobs/238077896) that matters to getting the tag added.

lcaflc commented:
Hi, yes I just see that yesterday night when tests failed.
I've started to figure out the issue and it seems related to an outdated setuptools version.
Will upate PR If I find the fix.

lcaflc committed:
Fix debian stretch build

bdist_wheel is now prodived by python-wheel module and no longuer
included into python-setuptools.

lcaflc commented:
It's now failing because systemd binary is not installed by default in base docker image.
In fact I don't really understand why you try to run the systemd init in the check instead of something a little simplier like /bin/true. Or run ansible itself to check that it work inside the build container.

Thanks.

lcaflc committed:
Add davedaemon as running process

Not the clever idea but replace the used systemd process by a simple sleeping daemon to keep the container alive when needed for tests.

lcaflc committed:
Revert "Add davedaemon as running process"

This reverts commit 3b03e4e751b45458ce1c7ad14eefe1ba69671e33.

lcaflc commented:
dockercloud check still fail, but I don't have access to status page to check why.

Thanks.

ChristopherDavenport commented on core/Debian/stretch/Dockerfile:
```diff
@@ -21,3 +22,6 @@ RUN apt-get update \
     && echo "localhost ansible_connection=local" >> /etc/ansible/hosts
+
+COPY davedaemon /usr/local/bin/
+RUN chmod 755 /usr/local/bin/davedaemon
```
I'm not sure how this davedaemon works. 

It seems as though without init there would be no way to test services?

lcaflc replied:
Hi,

Using travis-ci you currently bootup the docker image in background and expect it to run on a long term to then execute the tests using docker exec ... commands.

So you need a long term process to run this image without consuming all the cpu of the machine. This was traditionnaly done using the init process which just sleep for something to happen once it has started what needed. which is basically nothing inside a docker image.

However to save disk space and remove unused stuff by default, because an init system inside a docker image does not really make sense, systemd is not installed in debian stretch docker images now.

We can simply reinstall it in this image but it seems a little bit too much useless stuff for me. So I use this small dave fake daemon which is just a sleep loop.
I've tried with base binaries but didn't find something that works. (tried /bin/true, cat, ...) but they fail to wait for input because there is no tty to attach to. If you know another base binary which can just sleep for something to happen to keep the docker container running in background davedaemon is useless.

Thanks.

ChristopherDavenport replied:
Well the original reasoning behind the calls to systemd/init were that without that systems it becomes impossible to test services. Since this was originally attempting to mirror an OS build, not a simple docker container but using Docker containers to do this.

There are obvious difficulties with this plan. Primarily that using init it is practically impossible to contain services to inside their containers(specifically agetty which can be spawned as root outside the container and spike up to 100% cpu), however SystemD is better in this regard as using the setting it keeps services running with the proper containment. These are the difficulties associated with the docker running on the host kernel and the privileges associated with PID 1. 

So I am tenuous, as with these changes this would be unable to test services similar to a tomcat role that hooks itself into the service system. 

Other ideas are expanding to have a set of containers that use a minimal docker cmd arguments or a mini-service manage such as what you have done here so tests can be tailored to the environment as java or other binary installations generally don't require the heavy-handed service manager that is used currently in ansiblecheck.

What are your/general opinions, as I would like this to tailor generally? 

As you can see from the above there are some serious technical limitations I have been unable to engineer around at this point.

lcaflc replied:
Hi,
Very late reply but I have to make some tests to confirm your use case. I also add a similar issue in a recent project where the tested daemon does not want to start. At first I was looking at ansible failing to detect the proper init system to invoke the daemon startup but finish in something which was not so simple so I simply skip this in tests and didn't had time to look further.

However I don't see any case where the process started by the init system can go outside the container. This should never happen as container need to be isolated using cgroups and other low level primitives on the running system. If you have an exemple I'll be very happy to debug it :). But what can surely happen easily is that you can properly stop the container, when you stop the main entrypoint of the container but it's child you start using init system are no longuer proper child of it. You'll certainly need to destroy it in this case that will kill all process inside it.

Will get back shortly with updated patch, but I'll certainly just include systemd into the debian-stretch container.

lcaflc committed:
Add systemd-sysv

Add systemd-sysv package to debian-stretch container. Was removed from upstream debian base image but really needed for ansible tests using daemon services.

ChristopherDavenport reviewed (approved):
Looks good. As it doesn't look like your changes effected the failing build, I'll need to look into that.

Any thoughts?

lcaflc commented:
See issue #28, nothing related to this addition.

Thanks.

ChristopherDavenport commented:
Does this require a modification to the Docker Hub build script for stretch, not a big deal to implement. Just checking if this was updating or fully new functionality.

lcaflc commented:
I don't see your hub build configuration so I don't know if it needs updating. 
However cf: https://hub.docker.com/r/ansiblecheck/ansiblecheck/builds/ I see that a build is trigerred for debian-stretch tag, but not for debian-9 so certainly missing some part.

ChristopherDavenport commented:
Should build, as I just triggered the new tag. 👍
